### PR TITLE
[IMP] New guideline: avoid flaky tests

### PR DESCRIPTION
**Pull request by yajo**
This is a common problem across OCA repos.

I have done some research on the subject, including non-related projects, and also digging in upstream Odoo code to see how they deal with this.

Also I give some extra specific tips based on my own experience.

I hope this helps OCA have an official position about what and how to test.

@moduon

### CODE REVIEW

**Comment by yajo** [comment]
@sbidoul @moylop260 I guess you might be interested.

**Review by Shide — APPROVED** [review]

**Review by EmilioPascual — APPROVED** [review]

**Review by rousseldenis — COMMENTED** [review]
Everything is correct in this.

For the API part, @pedrobaeza API should not change ! Good APIs are implementing versioning to ensure the contract between the client (you) and the API remains the same. As @yajo explained in this, you are not responsible of API changes. Github test flows are not intended to detect such things, but ... to test your code and exclusively that.

As he said too, if you want to monitor, do it elsewhere, with another tool. As I already said, github actions are not launched periodically but manually. If nothing is done during 3 months but the API changes in that interval, you haven't detected the change (so your aim is not reached).

I think you should change your mind in front of growing people bothered by this (gentle remark).

**Comment by pedrobaeza** [comment]
I know that API shouldn't change, but they do, and without versioning. As you say, we are not responsible of them. A monitoring tool will only detect a healthcheck, but not specific API changes in the params and so on. Anyway, the point about `external` tag put at the end by Jairo seems fair to me. 

Please include the demo data point and I will approve it, and if you want to propose the switch to mock or tag as external in CTT, I will support it as well.

**Review by sbidoul — APPROVED** [review]
Great writeup @yajo. I'm fine with this PR.

I'd suggest mentioning [vcrpy](https://pypi.org/project/vcrpy-unittest/) as a alternative way to mock external API in a way that makes it easy to detect changes in the API (by simply running without the recorded cassettes).

**Review by legalsylvain — COMMENTED** [review]
hi. thanks for your contribution.

First, totally agree with the two first statements. 
- Beware with ``subTest`` pollution 
- Avoid dynamic dates 

**For external services**

> Remember that what you want to test is your code, not the external service.

Hi. I'm sorry but I'm not agree with that statement. When I develop tests, I want to be sure the module works. If it works with theoritical data mock but not with the current API in production, for me the module does'nt work, so the test does not fulfill its role.

> As I already said, github actions are not launched periodically but manually.

For the time being, in the OCA yes. But in fact, it is totally possible to create periodical test in github actions. 
i don't know how tag "external" works. Is it possible to run test excluding / including tests ? If yes, I'm in favor to say that : 
- call to external services _should_ be done in tests tagged with "external", and excluded from tests executed on each PR / each merge / etc.
- Add a new test executed once a month, that is running "external" tests. 

With such design : 
- New PR are not blocked if an existing module calls API that is broken
- We can ensure that the module works in the real life. (that is the most important point at the end) or at least be warned if something is broken.

**For demo data in tests**

There is pro and cons using demo data. So please don't add a blocking rule on that topic, and let developper choose what is the better way. (and also it is a matter of taste, at some point).

**General opinion**

In the last OCA days, there was a presentation (done by a board member I think, I don't remember who, sorry). He was saying that the number of newcomers that contribute once and then leave is growing. 

You have to be aware that adding more and more rules, linter and conventions accentuates this trend. (it's not the only cause of the problem, but it's one of them.)

At some point, the question is: do we prefer a very small community that respects all the ideal rules, or do we prefer a larger community with less stringent rules?

From my point of view, I think the OCA rules are already correct. (especially when I "leave" the OCA world and look at modules available on the odoo appstore, or modules installed by former integrators of new customers).

thanks.

File: website/Contribution/CONTRIBUTING.rst
Comment (luisg123v) [inline website/Contribution/CONTRIBUTING.rst:1111]: Maybe it's worth it to mention this library is already in Odoo's requirements, so a standard installation won't require installing any additional library.

**Review by luisg123v — CHANGES_REQUESTED** [review]

**Review by moylop260 — APPROVED** [review]
👍 

Thank you Yajo!

File: website/Contribution/CONTRIBUTING.rst
Comment (yajo) [reply website/Contribution/CONTRIBUTING.rst:1101]: I used 2 because the file is rst and the rest of it seems to be using 2. I don't think it's much important because it's just an example.

**Comment by Shide** [comment]
> Hi. I'm sorry but I'm not agree with that statement. When I develop tests, I want to be sure the module works. If it works with theoritical data mock but not with the current API in production, for me the module does'nt work, so the test does not fulfill its role.

You can't test production data every time a test runs (because there are no credentials). 
For this kind of cases, it's better to copy a response from production environment, anonymize it and put into a Mock without polluting the rest of the modules.

If an API changes something, you will see the error in production first because it's a daily use and you are not monitoring failing tests on GitHub. Then you should go to the code and fix whatever has changed. So, there is no point to test an API when running tests because nobody is going to discover the error before happens in production (API changes usually crashes something on Odoo or throws an error. IE.: You receive a list of categories in str format and after the API changes, you receive a list of IDs).

If nothing is using the module, someone install it and fails, can create an Issue or reach the maintainer to fix it soon.

If an error inside a running tests blocks another module because the service is down temporally, means that the test is not written properly (because you can't do nothing to fix anything).

Few APIs have endpoints to check the schema, so if your API has this endpoint, you can use it (catching exceptions related to down servers or something like this)

**Comment by pedrobaeza** [comment]
@yajo about the demo data, the principles are:

```
Demo data can be altered manually or by other third party module (through its fully XML-ID identifier), and thus, not having the expected data that are required for your tests to succeed.

Creating in the same test (or test suite) all - or at least most of - the data your test needs to be executed is the only guarantee to have a consistent and resilient result.
```

@legalsylvain about the question on entry barriers, I think these are recommendations that are very valuable when the whys are explained. They both serves to share knowledge between contributors, and to guide people to do good code. Not having them written makes that a lot of people not knowing them. It's also a reference that you can link when developing reasonings on reviews.

**Comment by yajo** [comment]
Thanks! I reworded it a bit, but the essence is the same.

Ready to merge on my side.

File: website/Contribution/CONTRIBUTING.rst
Comment (legalsylvain) [inline website/Contribution/CONTRIBUTING.rst:1242]:
```suggestion
If you create the test data within your test suite, you will have more consistent and resilient results.
```

reformulation so that it's advice, not an obligation.

**Review by legalsylvain — COMMENTED** [review]
minor remarks inline.
Otherwise LGTM. Thanks for this work.

> I think I worded the new guidelines with terms such as "avoid", "beware", "tips"... it's not like something written on stone, but at least we'll have a gathering place to explain why certain testing patterns are problematic.

That's the most important part for me. I'd like to avoid PR to be blocked, because reviewers ask extra works like duplicating demo data into test sections (for exemple). If it's not blocking point and just advices, it's great.

Thanks for all the documentation added.

**Comment by rvalyi** [comment]
Some practical feedback about demo data: For instance in OCA/l10n-brazil, fiscal settings are so complex that not only we need data for automated tests, but we also need data for advanced demo and advanced human tests... Maintaining all these demo data duplicated both for tests and demo would be kind of a nightmare...

At the end of the day what we try to do is: it's OK to reuse existing demo data for tests when it's master data (partners, products...). But it's not OK to reuse demo data for transactions as it breaks easily.
(and there are also many tests where we simply avoid demo data too).

Also we have some dedicated demo companies in the Brazilian root localization that we reuse extensively in the tests of the other localization modules.

Probably not something worth being part of the guidelines, but eventually this feedback helps.

**Comment by yajo** [comment]
I remember having thought in the past about reusing Odoo's xml-loading functions to re-load demo data under a controlled test environment. This way you avoid that kind of duplication, and also the uncontrolled input problem. However I didn't have time to materialize, and in the mean time Odoo decided to ship better common classes for tests (particularly [those of the account module](https://github.com/odoo/odoo/blob/06e0a005a3e96cd7708964498c996cbef358946c/addons/account/tests/common.py), which are very complex) so this problem became minor.

If we find some more tips to add, we can PR them later!

**Comment by rvalyi** [comment]
after writing this, I actually got the idea: in our case, we should probably put a few records creation (like companies, partners, products) in some tests/common.py method, call that in hooks.py before loading the demo XML files with demo transactions (is there any hook that can be called after pre_init and before loading demo data?). Thus we could reuse these controlled master data both in the demo data and in the isolated tests... Will report feedback...

**Comment by yajo** [comment]
> (is there any hook that can be called after pre_init and before loading demo data?)

Under a test case, you control the flow completely.

**Review by simahawk — APPROVED** [review]

**Comment by sbidoul** [comment]
Looks like all comments have been handled. Thanks for the work @yajo and the interesting conversation everyone.

**Pull request merged**
